mouse.lua

```diff
@@ -453,7 +454,117 @@ function mouse_event(typ, code, val)
   elseif code == 272 then
     mute = not mute
   end
-  
+  redraw()
+end
+
+-----------------------------------
+-- HID Keyboard Input
+-----------------------------------
+-- A S D F G - selects patterns 1 - 5
+-- 1 2 3 4   - voice 1-4 enable/disable
+-- SPACE     - toggles pattern playback
+-- TAB       - toggles melody/pairs mode
+-- ENTER     - random rate modulation
+-- J K L ;   - select clock division rates
+-- ↑ ↓ ← →   - transpose X/Y cursor 
+-- RIGHTSHIFT - toggle speed mod
+
+function keyboard.code(code,value)
+   if value == 1 or value == 2 then -- 1 is down, 2 is held, 0 is release
+    --voice enable toggles
+    if code == "1" then
+      toggle_bool_param("enables_" .. 1)
+    end
+    if code == "2" then
+      toggle_bool_param("enables_" .. 2)
+    end
+    if code == "3" then
+      toggle_bool_param("enables_" .. 3)
+    end
+    if code == "4" then
+      toggle_bool_param("enables_" .. 4)
+    end
+
+    --pattern select
+    if code == "A" then
+      -- Reset pattern indices.
+      params:set("pattern_index", 1)
+      pattern_counter_x = 1
+      pattern_counter_y = 1      
+    end
+    if code == "S" then
+      params:set("pattern_index", 2)
+      pattern_counter_x = 1
+      pattern_counter_y = 1
+    end
+    if code == "D" then
+      params:set("pattern_index", 3)
+      pattern_counter_x = 1
+      pattern_counter_y = 1
+    end
+    if code == "F" then
+      params:set("pattern_index", 4)
+      pattern_counter_x = 1
+      pattern_counter_y = 1
+    end
+    if code == "G" then
+      params:set("pattern_index", 5)
+      pattern_counter_x = 1
+      pattern_counter_y = 1
+    end
+    if code == "RIGHTSHIFT" then
+      toggle_bool_param("speed_mod")
+    end
+
+    --clock division rates
+    if code == "J" then
+      params:set("speed", 1)
+    end
+    if code == "K" then
+      params:set("speed", 2)
+    end
+    if code == "L" then
+      params:set("speed", 4)
+    end
+    if code == "SEMICOLON" then
+      params:set("speed", 8)
+    end
+    if code == "ENTER" then
+      params:set("speed", speeds[math.random(4)])
```

Comment:
I think `random(4)` could be `random(#speeds)` to make sure it stays up to date if any more speeds are added.